---
title: "Project Genome Analysis"
authors:
- Kaho Cheng
- Justine Claeys
- Willem Orye
- Lennert Verfaille
output:
  html_document: default
  pdf_document: default
subtitle: 'Transcription profiling by RNA-sequencing of normal and cirrhosis liver tissue and Kupffer cells' 
---
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(out.width = '60%')
knitr::opts_chunk$set(fig.width=12, fig.height=8)

# Data query
library(GEOquery)

# RNA-Seq analysis  
library(genefilter)
library(limma)
library(edgeR)
library(Glimma)
library(Rtsne)

# Annotation
library(annotate)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("EnsDb.Hsapiens.v75")
library("EnsDb.Hsapiens.v75")

# Data wrangling and plotting
library(gridExtra)
library(reshape2)
library(tidyverse)
library(factoextra)
```

# Introduction
Liver tissue and Kupffer cells from healthy patients and patients with liver cirrhosis were analysed. Both the cell types included 4 samples from patients with cirrhosis (cases) and 5 controls samples.

## Data collection (skip if you already have the DGElist file)

```{r}
# Download data
gse <- getGEO("GSE123661")
# Save experimental data
gsepheno <- pData(gse$GSE123661_series_matrix.txt.gz)
# Download supplementary files with raw count data
getGEOSuppFiles("GSE123661")
```

```{r}
# List the file names
files <- untar("./GSE123661/GSE123661_RAW.tar", list =T)
# Unzip tar file
untar("./GSE123661/GSE123661_RAW.tar")
# Read first five lines from the first file
read.delim(files[1], nrow = 5)
```

```{r}
#consolidate count data (count data contains duplicate rows)

filenames <- substring(files, 0, nchar(files)-3)
dir.create("GSE123661_noDup")

for (i in 1:length(files)){
  temp_df <- read.delim(files[i], stringsAsFactors = FALSE)
  temp_df.noDup <-aggregate(count~Gene.Symbol,data=temp_df,FUN=sum)
  write.table(temp_df.noDup, file = paste0("GSE123661_noDup/",filenames[i]), append = FALSE, sep = "\t", dec =".", row.names = FALSE, col.names = TRUE)
}

files.liver <- filenames[1:9]
files.kupffer <- filenames[10:18]

# Create DGElist obect (geneID & count)
dge.liver <- readDGE(paste0("GSE123661_noDup/",files.liver), columns=c(1,2))
dge.kupffer <- readDGE(paste0("GSE123661_noDup/",files.kupffer), columns=c(1,2))
str(dge.kupffer)
```

```{r}
# Add short samplenames
samplenames.liver <- paste0("L-", substring(colnames(dge.liver), 35,nchar(colnames(dge.liver))-7))
samplenames.kupffer <- paste0("K-", substring(colnames(dge.kupffer), 35,nchar(colnames(dge.kupffer))))

colnames(dge.liver)<-samplenames.liver
colnames(dge.kupffer)<-samplenames.kupffer

# Also create two seperate phenodata objects
gsepheno.liver <- gsepheno[which(gsepheno$source_name_ch1 %in% "Liver tissue"),]
gsepheno.kuppfer <- gsepheno[which(gsepheno$source_name_ch1 %in% "Macrophage"),]

# Add disease information to DGElist
dge.liver$samples$group <- gsub("disease state: ", "", as.character(gsepheno.liver$characteristics_ch1.1))
dge.kupffer$samples$group <- gsub("disease state: ", "", as.character(gsepheno.kuppfer$characteristics_ch1.1))

# Save DGElists
saveRDS(dge.liver, file='RawDGEList_GSE123661_Liver.rds')
saveRDS(dge.kupffer, file='RawDGEList_GSE123661_Kupffer.rds')

```

```{r}
# Annotation (optional, our data already contained the gene symbol)
symbol <- rownames(dge.liver)
genes <- AnnotationDbi::select(EnsDb.Hsapiens.v75,keys=symbol, columns=c("ENTREZID","GENENAME"), keytype="SYMBOL")
columns(EnsDb.Hsapiens.v75)

genes<-genes[!duplicated(genes$ENTREZID),]
dge.liver$genes <- genes
dge.kupffer$genes <- genes
```

## Data exploration (start here if DGElist objects are available)

```{r}
dge.liver <- readRDS('RawDGEList_GSE123661_Liver.rds')
dge.kupffer <- readRDS('RawDGEList_GSE123661_Kupffer.rds')
```

```{r}
# Keep copy of the raw counts
liver.raw <- dge.liver
kupffer.raw <- dge.kupffer
```

Plot the raw library sizes. 

```{r}
#liver
liver.raw$samples["sample"] <- colnames(dge.liver)
ggplot(data=liver.raw$samples, aes(x = sample, y = lib.size)) + 
  geom_bar(stat="identity", aes(fill = group))+
  labs(title = "Raw Library Sizes; Liver", x = "Sample", y = "Library Size")

#kupffer
kupffer.raw$samples["sample"] <- colnames(dge.kupffer)
ggplot(data=kupffer.raw$samples, aes(x = sample, y = lib.size)) + 
  geom_bar(stat="identity", aes(fill = group))+
  labs(title = "Raw Library Sizes; Kupffer", x = "Sample", y = "Library Size")
```

    
# Preprocessing

## Counts per Million (CPM)

```{r}
# Converting to CPM
cpm.liver <- cpm(dge.liver) 
cpm.kupffer <- cpm(dge.kupffer)
head(cpm.liver[, 1:4])
```

```{r}
# Calculate log counts per million values
lcpm.liver <- cpm(dge.liver, log=TRUE)
lcpm.kupffer <- cpm(dge.kupffer, log=TRUE)
head(lcpm.liver[, 1:4])
```

## Raw data filtering

Next, we will filter out genes with low cpm values. Genes that are not expressed in any sample will eventually not result in differential expression. Additionally, genes with low or zero cpm values in most of the samples are unlikely to result in differential expression. It is best to remove these genes as these genes will also influence the outcome for multiple testing (= more power after removing them).

Determine how many genes show no expression across all samples?

```{r}
table(rowSums(dge.liver$counts==0)==9) #9 samples (4 cases, 5 controls) #20816 FALSE; 11 TRUE
table(rowSums(dge.kupffer$counts==0)==9) #9 samples (4 cases, 5 controls) #20825 FALSE; 2 true
```

Remove all genes that do not have a cpm value of at least 0.5 in at least three samples:
(A cpm value of 0.5 corresponds to 10 reads for a library size of 20.000.000.)

```{r}
# in total 20827 genes (interesting + uninteresting genes)
liver.keep.exprs3 <- filterByExpr(dge.liver, group=dge.liver$sampples$group) #smallest group is 4
sum(liver.keep.exprs3) #15051 genes left

kupffer.keep.exprs3 <- filterByExpr(dge.kupffer, group=dge.kupffer$sampples$group) #smallest group is 4
sum(kupffer.keep.exprs3) #16996 genes left
```

```{r}
dge.liver <- dge.liver[liver.keep.exprs3,, keep.lib.sizes=FALSE]
dge.kupffer <- dge.kupffer[kupffer.keep.exprs3,, keep.lib.sizes=FALSE]
```

Let's compare the raw data with the filtered data using density plots:

```{r}
#liver
lcpm.raw.counts.L <- cpm(liver.raw, log=TRUE, normalized.lib.sizes = FALSE)
lcpm.raw.counts.plot.L <- melt(as.data.frame(lcpm.raw.counts.L)) %>% 
  rename("Sample" = variable, "Expression" = value) 

ggplot(lcpm.raw.counts.plot.L, aes(x = Expression, colour = Sample)) + 
  geom_line(stat = 'density')+
  labs(title = 'Raw data Liver',  x = 'Log-cpm')

lcpm.filtered.counts.L <- cpm(dge.liver, log=TRUE, normalized.lib.sizes = FALSE)
lcpm.filtered.counts.plot.L <- melt(as.data.frame(lcpm.filtered.counts.L))  %>%
  rename("Sample" = variable, "Expression" = value) 

ggplot(lcpm.filtered.counts.plot.L, aes(x = Expression, colour = Sample)) + 
  geom_line(stat = 'density')+
  labs(title = 'Filtered data Liver',  x = 'Log-cpm')

#kupffer
lcpm.raw.counts.K <- cpm(kupffer.raw, log=TRUE, normalized.lib.sizes = FALSE)
lcpm.raw.counts.plot.K <- melt(as.data.frame(lcpm.raw.counts.K)) %>% 
  rename("Sample" = variable, "Expression" = value) 

ggplot(lcpm.raw.counts.plot.K, aes(x = Expression, colour = Sample)) + 
  geom_line(stat = 'density')+
  labs(title = 'Raw data Kupffer',  x = 'Log-cpm')

lcpm.filtered.counts.K <- cpm(dge.kupffer, log=TRUE, normalized.lib.sizes = FALSE)
lcpm.filtered.counts.plot.K <- melt(as.data.frame(lcpm.filtered.counts.K))  %>%
  rename("Sample" = variable, "Expression" = value) 

ggplot(lcpm.filtered.counts.plot.K, aes(x = Expression, colour = Sample)) + 
  geom_line(stat = 'density')+
  labs(title = 'Filtered data Kupffer',  x = 'Log-cpm')
```

## Normalisation

Normalisation is a process designed to identify and remove systematic technical differences between samples that occur in the data to ensure that technical bias has minimal impact on the results.

### Trimmed Mean of M-values (TMM) (Robinson and Oshlack 2010)

Using the `calcNormFactors` function from the `edgeR` library, we normalise our data using TMM normalisation:

```{r}
# Keep copy of non-normalised data
dge.liver.before.norm <- dge.liver
dge.kupffer.before.norm <- dge.kupffer

# Perform TMM normalisation
dge.liver <- calcNormFactors(dge.liver, method = "TMM")
dge.liver$samples['sample_id'] <- rownames(dge.liver$samples)
dge.liver$samples$norm.factors

dge.kupffer <- calcNormFactors(dge.kupffer, method = "TMM")
dge.kupffer$samples['sample_id'] <- rownames(dge.kupffer$samples)
dge.kupffer$samples$norm.factors
```
````{r}
#save preprocessed DGElist objects
saveRDS(dge.liver, file = "Preprocessed_DGEList_GSE123661_Liver.rds")
saveRDS(dge.kupffer, file = "Preprocessed_DGEList_GSE123661_Kupffer.rds")
````

Have a look at the effect of normalisation by making boxplots of log cpm values before and after normalisation. 

```{r}
#kupffer
lcpm.K <- edgeR::cpm(dge.kupffer.before.norm, log=TRUE)
lcpm.plot.K <- melt(as.data.frame(lcpm.K)) %>% rename("Sample" = variable,
"Expression" = value)
ggplot(lcpm.plot.K, aes(y= Expression, colour = Sample)) + geom_boxplot() +
labs(title="Kupffer: CPM with Unnormalised LS",y="Log-cpm")
lcpm.norm.K <- edgeR::cpm(dge.kupffer, log=TRUE)
lcpm.norm.plot.K <- melt(as.data.frame(lcpm.norm.K)) %>%
rename("Sample" = variable, "Expression" = value)
ggplot(lcpm.norm.plot.K, aes(y= Expression, colour = Sample)) + geom_boxplot() +
labs(title="Kupffer: CPM with Normalised LS",y="Log-cpm")

#liver
lcpm.L <- edgeR::cpm(dge.liver.before.norm, log=TRUE)
lcpm.plot.L <- melt(as.data.frame(lcpm.L)) %>% rename("Sample" = variable,
"Expression" = value)
ggplot(lcpm.plot.L, aes(y= Expression, colour = Sample)) + geom_boxplot() +
labs(title="Liver: CPM with Unnormalised LS",y="Log-cpm")
lcpm.norm.L <- edgeR::cpm(dge.liver, log=TRUE)
lcpm.norm.plot.L <- melt(as.data.frame(lcpm.norm.L)) %>%
rename("Sample" = variable, "Expression" = value)
ggplot(lcpm.norm.plot.L, aes(y= Expression, colour = Sample)) + geom_boxplot() +
labs(title="Liver: CPM with Normalised LS",y="Log-cpm")

```

There is only a minor effect

The effective library size can be obtained by multiplying the library size with additional normalisation factors. Note that for the most count-based statistical testing procedures, the normalization factors are used directly in the statistical models. In other words, in most cases you can simply supply methods such as EdgeR with the raw count data and the scaling factors.

```{r}
effSize.liver = dge.liver$samples$lib.size * dge.liver$samples$norm.factors
effSize.kupffer = dge.kupffer$samples$lib.size * dge.kupffer$samples$norm.factors
```

# Dimensionality reduction
Genomic datasets are often high-dimensional. Dimensionality reduction analyses provide a way of reducing the complex dataset into a lower dimension (that is more accessible for interpretation) while attempting to retain as much relevant information as possible. 

## MDS
The `plotMDS` function performs dimensionality reduction and the distances between the gene expression profiles are visualised. The distances correspond to the leading **log-fold-change** (if the input data is in the log scale), determined as the root mean square (Euclidean) average of the largest log2-fold changes between each **pair** of samples (the standard option looks at the top 500 genes with the largest log-fold change for each pair). 

```{r}
# The underlying MDS function is the cmdscale function
# We need the log transformed lcpm counts
mds.liver <- plotMDS(lcpm.norm.L, dim = c(1,8), gene.selection = 'pairwise', plot = FALSE)
mds.kupffer <- plotMDS(lcpm.norm.K, dim = c(1,8), gene.selection = 'pairwise', plot = FALSE)
```

```{r}
ggplotMDS <- function(mds, metadata, param, dim1, dim2, label){
  
  # Create MDS plot with ggplot
  
  # Input: mds object
  #       annotation dataframe
  #       parameter of interest in the annotation dataframe 
  #       first dimension of interest
  #       second dimension of interest
  #       how you want the samples to be labeled

  # Extract cmdscale from mds object
  dataMDS <- as.data.frame(mds$cmdscale.out)
  
  i <-  metadata[colnames(metadata) == param]
  l <-  metadata[colnames(metadata) == label]
  
  plot <- ggplot(data = dataMDS, aes(x = dataMDS[,dim1] , 
                                     y = dataMDS[,dim2])) +
    geom_point(aes(colour = as.factor(i[,1])), size = 2)+
    labs(x = paste0("Dimension ", dim1), y = paste0("Dimension ", dim2), 
         colour = param) +
    geom_text(aes(label = l[,1], colour = as.factor(i[,1])), 
              size = 2.8, vjust = -0.5) +
    theme(axis.title.x=element_text(size=11),
          axis.title.y=element_text(size=11),
          legend.text = element_text(size=10), 
          legend.title = element_text(size = 11, face = "bold"))
  
return(plot)
}
```

```{r}
#liver
ggplotMDS(mds.liver, as.data.frame(dge.liver$samples), 'group', 1, 2, 'sample_id')
ggplotMDS(mds.liver, as.data.frame(dge.liver$samples), 'group', 1, 3, 'sample_id')
ggplotMDS(mds.liver, as.data.frame(dge.liver$samples), 'group', 3, 4, 'sample_id')

#kupffer (by looking at MDS plots differences are less pronounced here)
ggplotMDS(mds.kupffer, as.data.frame(dge.kupffer$samples), 'group', 1, 2, 'sample_id')
ggplotMDS(mds.kupffer, as.data.frame(dge.kupffer$samples), 'group', 1, 3, 'sample_id')
ggplotMDS(mds.kupffer, as.data.frame(dge.kupffer$samples), 'group', 3, 4, 'sample_id')
```

An interactive MDS plot can be created with the `glimma` package. 

```{r}
#liver
#glMDSPlot(lcpm.norm.L,labels=paste(dge.liver$samples$group,sep="_"), groups=dge.liver$samples[,c(2,5)],launch=T)
```

```{r}
#kupffer
#glMDSPlot(lcpm.norm.K,labels=paste(dge.kupffer$samples$group,sep="_"), groups=dge.kupffer$samples[,c(2,5)],launch=T)
```

We can see that the samples from the kupffer cells are more spread out and there is less separation between the control and case samples. So we expect less significant differential expression compared to the liver tissue samples.

#differential expression using EdgeR

## Load the Data (if needed)

```{r}
dge.liver <- readRDS('Preprocessed_DGEList_GSE123661_Liver.rds')
dge.kupffer <- readRDS('Preprocessed_DGEList_GSE123661_Kupffer.rds')
```

The goal: compare two cell populations (cirrhosis vs control) --> Are there any genes differentially expressed between the different cell populations?

## Specify the Design Matrix

```{r}
#liver
dge.liver$samples$group <- as.factor(dge.liver$samples$group)
design_liver <- model.matrix(~0+group, data = dge.liver$samples)
colnames(design_liver) <- levels(dge.liver$samples$group)
design_liver
#kupffer
dge.kupffer$samples$group <- as.factor(dge.kupffer$samples$group)
design_kupffer <- model.matrix(~0+group, data = dge.kupffer$samples)
colnames(design_kupffer) <- levels(dge.kupffer$samples$group)
design_kupffer
```

## Estimate Dispersion and Fit Model

We fit genewise generalized linear models. 

```{r}
dge.liver <- estimateDisp(dge.liver, design = design_liver)
names(dge.liver)
dge.kupffer <- estimateDisp(dge.kupffer, design = design_kupffer)
```

###BCV plot

Create a BCV plot to visualise the dispersion estimates. 
```{r}
plotBCV(dge.liver)
plotBCV(dge.kupffer)
```


### Fit the model 

```{r}
fit.kupffer <- glmQLFit(dge.kupffer, design_kupffer)
fit.liver <- glmQLFit(dge.liver, design_liver)

names(fit.liver)
```

### QLD

Plot the QL dispersions. 
```{r}
plotQLDisp(fit.liver) # Plot the quasi-likelihood dispersion 
plotQLDisp(fit.kupffer)
```


## Test Differential Expression using glmQLFit

Make a contrast matrix to specify comparisons:

```{r}
cont.matrix_L <- makeContrasts("Cirrhosis-Control", levels=design_liver) 
cont.matrix_K <- makeContrasts("Cirrhosis-Control", levels=design_kupffer) 
```

Conduct quasi-likelihood tests to assess differential expression :

```{r}
qlf.kupffer <- glmQLFTest(fit.kupffer, contrast = cont.matrix_K[ ,1])
qlf.liver <- glmQLFTest(fit.liver, contrast = cont.matrix_L[ , 1]) # 
```

Determine how many genes have an FDR-rate lower than 0.01?

```{r}
# Number of genes FDR < 0.01

#liver
top.qlf.liver <- topTags(qlf.liver, n=Inf)
sum(top.qlf.liver$table$FDR<0.01) #34 rather low for GSA (ideally 100-500 DE genes), for GSA use genes with FDR >0.05 -> 405 genes
head(top.qlf.liver)

#kupffer
top.qlf.kupffer <- topTags(qlf.kupffer, n=Inf)
sum(top.qlf.kupffer$table$FDR<0.01) # <0.01 -> 0 genes (Top gene FDR = 0.368) already expected less DE genes compared to the liver cells when we saw the  MDS plots
head(top.qlf.kupffer)
```

conclusion:
The liver tissue has 34 significantly DE genes with an FDR < 0.01 --> with GSA we can find out which pathways are over represented. We will use DE genes with an FDR < 0.05 to include enough DE genes (405 genes). For GSA this higher FDR will not be a problem.
Kupffer cells show no DE genes (top gene FDR = 0.368). We already expected less DE genes compared to the liver cells when we saw the  MDS plots. From here we only work further with the liver tissue cells.


## Accounting for confounders
The phenodata didn't mention any confounders and their research is not available.
So unfortunately we could not take into account the effect of confounders like age, gender, alcohol use, etc. on the relation between disease and gene expression.

#save EdgeR DE-output for next step:

````{r}
#save data needed for GSA as text
write.table(top.qlf.liver$table,"EdgeR_output_Liver.txt",sep="\t",row.names=T)
write.table(top.qlf.kupffer$table,"EdgeR_output_Kupffer.txt",sep="\t",row.names=T) #actually not needed
````

# Gene set analysis

"overrepresentation analysis": certain GO-term (Gene Ontology term) or a certain pathway is overrepresented amongst our set of interest (e.g. differentially expressed genes) compared to the background. The selection of both this set of interest and the background is of vital importance. Note that selecting a faulty background can already introduce (some) overrepresentation. 

To test for overrepresentation, for each term of interest (GO/Pathway) the number of genes in our set of interest and in the background is counted and then tested (typically by chi-squared/Fisher exact test) whether the occurence in the set of interest is significantly different from the occurence in the background. We then correct for multiple testing based on the number of gene sets we analyse.

Load in the data (output DE by EdgeR,)

```{r}
res.liver <- read.table("./EdgeR_output_Liver.txt", header = T, row.names = NULL, quote="", sep="\t")
colnames(res.liver) <- c("Gene.symbol","logFC","logCPM","F","Pvalue","FDR")
```

A FDR cut off of 0.01 gives 34 DE genes. Therefore we chose a FDR cut off of 0.05 which will include 405 DE genes in our GSA set.
Once we have have our background genes and gene set of interest saved as a textfile, we can use the webtool Webgestalt to perform GSA.

```{r}
# Select Gene symbols for the background (i.e. all expressed genes)
Gene.symbol.background <- res.liver$Gene.symbol
Gene.symbol.background <- gsub("\"","",Gene.symbol.background)

write.table(Gene.symbol.background, file="Gene_symbol_background.txt", 
            col.names=F, quote=F, row.names=F)

# Select Gene symbols for the set of interest
Gene.symbol.sign <- res.liver$Gene.symbol[(res.liver$FDR<0.05)]
Gene.symbol.sign <- Gene.symbol.sign[!is.na(Gene.symbol.sign)]
Gene.symbol.sign <- sort(as.character(Gene.symbol.sign))
length(Gene.symbol.sign)
Gene.symbol.sign <- gsub("\"","",Gene.symbol.sign)

write.table(Gene.symbol.sign, file="Gene_symbol_sign.txt", 
            col.names=F, quote=F, row.names=F)
```

Now that we have have our background genes and gene set of interest saved as a textfile, we can use the webtool Webgestalt to perform GSA.
The geneontology database was selected to find out which biological processes are overrepresented in our gene set of interest.

```{r}
htmltools::includeHTML('Report_wg_result1620406440.html')
```

